src/services/activeUsers.js

```diff
@@ -0,0 +1,124 @@
+    stream.end();
+  };
+  query.on('row', (r) => {
+    if (writeHeader) {
+      writeHeader = false;
+      const header = Object.keys(r).toString();
+      stream.push(`${header}\n`);
+    }
+    stream.push(transform ? transform(r) : r);
+  });
+  query.once('error', (err) => {
+    stream.emit('error', err);
+    end();
+  });
+  query.once('end', end);
+
+  return stream;
+};
+
+/**
+ * Retrieves users that have permissions.
+ * Inspired by https://github.com/sequelize/sequelize/issues/2454
+ *
+ * @param {*} sql - raw sql
+ * @param {*} transform - a function to transform a row of data
+ * @returns {Promise<Object>} - an Active Users stream
+ */
+export default async function activeUsers() {
+  const sql = `WITH "ActiveUsers" AS (
+      SELECT
+          "User"."id" AS "id",
+          "User"."name" AS "name",
+          "User"."email" AS "email",
+          "User"."homeRegionId" AS "homeRegionId",
+          "User"."lastLogin" AS "lastLogin",
+          "User"."flags" AS "flags",
+          "roles"."roleId" AS "roleId",
+          "Role".name AS "roles",
+          "permissions"."scopeId" AS "scopeId",
+          "Scope"."name" AS "scope",
+          "permissions"."regionId" AS "regionId"
+      FROM
+          "Users" AS "User"
+          INNER JOIN "Permissions" AS "permissions" ON "User"."id" = "permissions"."userId"
+          LEFT JOIN "UserRoles" "roles" ON "roles"."userId" = "User"."id"
+          LEFT JOIN "Roles" "Role" ON "roles"."roleId" = "Role"."id"
+          LEFT JOIN "Scopes" "Scope" ON "Scope"."id" = "permissions"."scopeId"
+  )
+  SELECT
+      "ActiveUsers"."name" AS "Name",
+      "ActiveUsers"."email" AS "Email",
+      "ActiveUsers"."homeRegionId" AS "Region",
+      string_agg(DISTINCT "ActiveUsers"."roles" :: VARCHAR, ',') AS "Roles",
+      "ActiveUsers"."lastLogin" AS "Last login",
+      MIN(CASE WHEN "ActiveUsers"."scope" = 'SITE_ACCESS' THEN "ActiveUsers"."scope" END) AS "SITE_ACCESS",
+      MIN(CASE WHEN "ActiveUsers"."scope" = 'ADMIN' THEN "ActiveUsers"."scope" END) AS "ADMIN",
+      MIN(CASE WHEN "ActiveUsers"."scope" = 'UNLOCK_APPROVED_REPORTS' THEN "ActiveUsers"."scope" END) AS "UNLOCK_APPROVED_REPORTS",
+      MIN(CASE WHEN "ActiveUsers"."scope" = 'READ_WRITE_REPORTS' THEN "ActiveUsers"."scope" END) AS "READ_WRITE_REPORTS",
+      array_agg(DISTINCT "ActiveUsers"."regionId") filter (WHERE "ActiveUsers"."scope" = 'READ_WRITE_REPORTS') AS "READ_WRITE_REPORTS regions",
+      MIN(CASE WHEN "ActiveUsers"."scope" = 'READ_REPORTS' THEN "ActiveUsers"."scope" END) AS "READ_REPORTS",
+      array_agg(DISTINCT "ActiveUsers"."regionId") filter (WHERE "ActiveUsers"."scope" = 'READ_REPORTS') AS "READ_REPORTS regions",
+      MIN(CASE WHEN "ActiveUsers"."scope" = 'APPROVE_REPORTS' THEN "ActiveUsers"."scope" END) AS "APPROVE_REPORTS",
+      array_agg(DISTINCT "ActiveUsers"."regionId") filter (WHERE "ActiveUsers"."scope" = 'APPROVE_REPORTS') AS "APPROVE_REPORTS regions",
+      "ActiveUsers"."flags" AS "Flags"
```

Comment:
https://blog.jooq.org/the-performance-impact-of-sqls-filter-clause/
CASE is slower then filter, so the MIN(CASE columns would be better as string_agg(DISTINCT) filter

Reply:
Interesting. Initially, I changed to using string agg and filter. The case ended up staying because the requirements changed slightly.

Reply:
In your implementation they are effectively the same output, just the min case is slower.

Reply:
I actually didn't see any consistent performance improvement on our small dataset when running on my local machine. Perhaps this would be more visible with larger results, though.  In any case, since the requirements changed slightly, we are returning different data now, a string "Yes" rather than the permission name.